proposals/0010-resource-element-type-validation.md

```diff
@@ -0,0 +1,217 @@
+Below is a description of resources that are considered "typed buffers" vs "raw buffers".
+* typed buffers
+  * [RW|RasterizerOrdered]Buffer
+  * [Feedback]Texture*
+* raw buffers
+  * [Append|Consume|RW|RasterizerOrdered]StructuredBuffer
+  * [RW]ByteAddressBuffer
+
+There is a distinct set of rules that define valid element types for typed buffer resources
+and valid element types for raw buffer resources.
+
+Element types for typed buffer resources:
+* Are not intangible (e.g., isn't a resource type)
+* Must be vectors or scalars of arithmetic types, not bools nor enums nor arrays
+* Should be a scalar or homogenous vector of a floating-point or integer type, with a maximum of 4 components after translating 64-bit components into pairs of uint32_t components
+Element types for raw buffer resources:
+* Are not intangible (e.g., isn't a resource type)
+* All constituent types must be arithmetic types or bools or enums
+
+Resource types are never allowed as element types (i.e., `RWBuffer<int>` as an element type).
+If someone writes `RWBuffer<MyCustomType>` and MyCustomType is not a valid element type, 
+there should be infrastructure to reject this element type and emit a message explaining 
+why it was rejected as an element type.
+
+## Motivation
+Currently, there is an allow list of valid element types where element types not on the list
+are rejected. It must be modified with respect to this spec. The allow list isn't
+broad enough, because user-defined types aren't allowed for raw buffer resources.
+Ideally, a user should be able to determine exactly how any user-defined type is invalid 
+as an element type. Some system should be in place to more completely enforce the rules for 
+valid and invalid element types, as well as provide useful information on why they are invalid.
+
+For example, `RWBuffer<double4> b : register(u4);` will emit an error in DXC, but will not in 
+clang-dxc, despite the fact that `double4` is an invalid element type for typed buffers.
+
+## Proposed solution
+
+The proposed solution is to modify the declaration of each resource declared in 
+`clang\lib\Sema\HLSLExternalSemaSource.cpp` and insert into each representative
+AST node a concept. The AST node will be created as if the C++20 `concept` keyword
+was parsed and applied to the declaration. The concept will be used to validate the
+given element type, and will emit errors when the given element type is invalid. 
+Although concepts are not currently supported in HLSL, we expect support to be 
+added at some point in the future. Meanwhile, because LLVM does support concepts, 
+we can make use of them when constructing the AST in Sema.
+
+Two builtins will be used to validate typed buffer element types. Any resource 
+element type may not be intangible, so the negation of `__builtin_hlsl_is_intangible`
+will be used for both typed and raw buffer resources.
+A new built-in, `__builtin_hlsl_typed_element_compatible`, will be added in order

```

Comment:
Not to bike shed, but this builtin name seems odd to me because it doesn't really describe what this is an element of. Maybe `__builtin_hlsl_typed_buffer_element_compatible` or `__builtin_hlsl_typed_resource_element_compatible` to add the missing context?

Reply:
Sounds good, I'll go with typed resource.